### PR TITLE
feat: integrate deepgemm into EPMoE

### PR DESCRIPTION
**Pull request by xutizhou**
## Motivation

For normal EPMoE (no DeepEP), integrate DeepGEMM as an option.

This PR builds upon the work from [pr5805](https://github.com/sgl-project/sglang/pull/5805) Credit goes to @TianQiLin666666, who authored most of the code presented here.
```
Co-authored-by: @TianQiLin666666
```
## Performance
two h20 node 

node 1
```
python -m sglang.launch_server --model-path /path/to/DeepSeek-V3/    --trust-remote-code --tp-size 16 --enable-dp-attention --dp-size 16 --mem-fraction-static 0.8 --enable-ep-moe --disable-radix-cache --dist-init-addr 10.6.131.5:5000 --nnodes 2 --node-rank 0
```

node 2
```
python -m sglang.launch_server --model-path /path/to/DeepSeek-V3/    --trust-remote-code --tp-size 16 --enable-dp-attention --dp-size 16 --mem-fraction-static 0.8 --enable-ep-moe --disable-radix-cache  --dist-init-addr 10.6.131.5:5000 --nnodes 2 --node-rank 1
```

test
```
python3 -m sglang.bench_one_batch_server --model /path/to/DeepSeek-V3/ --base-url http://localhost:30000 --batch-size 1 16 32 64 128 --input-len 1024 --output-len 1024
```
| Batch Size | Test Group | Output Throughput (tok/s) |
|------------|---------------|---------------------------|
| 1 | w/o deepgemm | 24.95 |
| | w deepgemm | 30.49 |
| | Diff | +22.2% |
| 16 | w/o deepgemm | 309.32 |
| | w deepgemm | 335.31 |
| | Diff | +8.4% |
| 32 | w/o deepgemm | 553.30 |
| | w deepgemm | 595.04 |
| | Diff | +7.5% |
| 64 | w/o deepgemm | 1003.07 |
| | w deepgemm | 1088.89 |
| | Diff | +8.6% |
| 128 | w/o deepgemm | 1644.28 |
| | w deepgemm | 1786.10 |
| | Diff | +8.6% |

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 1]
```diff
@@ -228,13 +238,140 @@ def __init__(
...
+        num_groups, m, k = gateup_input_fp8[0].size()
+        n = self.w13_weight.size(1)
+        gateup_output = torch.empty(
+            (num_groups, m, n), device=hidden_states_device, dtype=torch.bfloat16
+        )
+
+        m_grouped_gemm_fp8_fp8_bf16_nt_masked(
+            gateup_input_fp8, self.w13_weight_fp8, gateup_output, masked_m, expected_m
+        )
+        del gateup_input_fp8
```
Comment (fzyzcjy) [inline python/sglang/srt/layers/moe/ep_moe/layer.py:338]: this may not work b/c gateup_input still has nonzero ref count

**Comment by TianQiLin666666** [comment]
### Compare performance of DeepGEMM with triton groupgemm
on 8 * H20-96G
* DeepGEMM
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 8
Successful requests:                     8
Benchmark duration (s):                  58.65
Total input tokens:                      28000
Total generated tokens:                  12000
Total generated tokens (retokenized):    11924
Request throughput (req/s):              0.14
Input token throughput (tok/s):          477.41
Output token throughput (tok/s):         204.61
Total token throughput (tok/s):          682.02
Concurrency:                             7.99
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   58606.13
Median E2E Latency (ms):                 58610.62
---------------Time to First Token----------------
Mean TTFT (ms):                          3021.35
Median TTFT (ms):                        3106.69
P99 TTFT (ms):                           4465.86
---------------Inter-Token Latency----------------
Mean ITL (ms):                           37.08
Median ITL (ms):                         35.99
P95 ITL (ms):                            37.48
P99 ITL (ms):                            38.09
Max ITL (ms):                            3404.95
==================================================
```
* triton groupgemm
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 8
Successful requests:                     8
Benchmark duration (s):                  69.08
Total input tokens:                      28000
Total generated tokens:                  12000
Total generated tokens (retokenized):    11950
Request throughput (req/s):              0.12
Input token throughput (tok/s):          405.33
Output token throughput (tok/s):         173.71
Total token throughput (tok/s):          579.04
Concurrency:                             7.99
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   69027.51
Median E2E Latency (ms):                 69032.62
---------------Time to First Token----------------
Mean TTFT (ms):                          3320.52
Median TTFT (ms):                        3502.54
P99 TTFT (ms):                           4823.64
---------------Inter-Token Latency----------------
Mean ITL (ms):                           43.83
Median ITL (ms):                         42.81
P95 ITL (ms):                            43.81
P99 ITL (ms):                            44.26
Max ITL (ms):                            3701.18
==================================================
```

**Review by zyksir — COMMENTED** [review]
@TianQiLin666666 amazing! Did you test the correctness of your implementation? like using mmlu or gsm8k using baseline and deepgemm?

**Comment by TianQiLin666666** [comment]
> @TianQiLin666666 amazing! Did you test the correctness of your implementation? like using mmlu or gsm8k using baseline and deepgemm?

Yes, I have tested mmlu in https://github.com/sgl-project/sglang/pull/5805. And I will test again before it is merged.

File: python/sglang/srt/layers/moe/ep_moe/kernels.py [hunk 2]
```diff
@@ -1085,3 +1085,207 @@ def tma_align_input_scale(input_scale: torch.Tensor):
...
+
+    for idx in range(topk):
+        expert_id = tl.load(topk_ids_ptr + idx)
+        if expert_id >= start_expert_id and expert_id <= end_expert_id:
+            dst_idx_int32 = tl.load(src2dst_ptr + idx)
+            dst_idx = dst_idx_int32.to(tl.int64)
+            dst_idx = dst_idx - start_expert_id * m_max
+            dst_ptr = gateup_input_ptr + dst_idx * hidden_size
+            for start_offset in tl.range(0, hidden_size, BLOCK_SIZE):
+                offset = start_offset + tl.arange(0, BLOCK_SIZE)
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/kernels.py]: If we pre-define a tl.arange() outside this for-loop and reuse it, can this kernel be faster?

File: python/sglang/srt/layers/moe/ep_moe/kernels.py [hunk 3]
```diff
@@ -1085,3 +1085,207 @@ def tma_align_input_scale(input_scale: torch.Tensor):
...
+        masked_m[start_expert_id : (end_expert_id + 1)],
+        expected_m,
+        src2dst,
+        gateup_input,
+        gateup_input_scale,
+    )
+
+
+@triton.jit
+def deepgemm_post_reorder_triton_kernel(
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/kernels.py]: this kernel is very similar with `post_reorder_triton_kernel`. Can we simply add an additional argument to `post_reorder_triton_kernel` to control the workflow? This can avoid repetitive implementation.

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 4]
```diff
@@ -232,13 +242,144 @@ def __init__(
...
         hidden_states_dtype = hidden_states.dtype
         hidden_states_device = hidden_states.device
+        topk_weights, topk_ids = select_experts(
+            hidden_states=hidden_states,
+            router_logits=router_logits,
+            top_k=self.top_k,
+            use_grouped_topk=self.use_grouped_topk,
+            renormalize=self.renormalize,
+            topk_group=self.topk_group,
+            num_expert_group=self.num_expert_group,
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/layer.py:266]: we recently added `num_fused_shared_experts`

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 5]
```diff
@@ -232,13 +242,144 @@ def __init__(
...
     def forward(self, hidden_states: torch.Tensor, router_logits: torch.Tensor):
+        if use_deep_gemm and _ENABLE_JIT_DEEPGEMM and self.use_fp8_w8a8:
+            return self.forward_deepgemm(hidden_states, router_logits)
+        else:
+            return self.forward_normal(hidden_states, router_logits)
+
+    def forward_deepgemm(
+        self, hidden_states: torch.Tensor, router_logits: torch.Tensor
+    ):
+        assert self.quant_method is not None
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/layer.py:254]: assert self.activation == "silu"

Comment (TianQiLin666666) [reply python/sglang/srt/layers/moe/ep_moe/layer.py:254]: Done!

File: python/sglang/srt/layers/moe/ep_moe/kernels.py (hunk 3 above)
Comment (TianQiLin666666) [reply python/sglang/srt/layers/moe/ep_moe/kernels.py]: Done!

File: python/sglang/srt/layers/moe/ep_moe/kernels.py (hunk 2 above)
Comment (TianQiLin666666) [reply python/sglang/srt/layers/moe/ep_moe/kernels.py]: Done!

**Comment by TianQiLin666666** [comment]
## Accuracy
* gsm8k
```
python3 bench_sglang.py --num-questions 320 --parallel 16 --data-path /data/datasets/gsm8k/test.jsonl --port 8080
```
* mmlu
```
python3 bench_sglang.py --nsub 50 --port 8080 --data_dir /data/datasets/mmlu
```

|    | MMLU | GSM8K |
| --- | --- | ---|
|main branch | 0.809 | 0.909 |
| this PR | 0.809 | 0.909 |

**Review by ch-wan — APPROVED** [review]
